FWCore/Framework/interface/GenericProduct.h

```diff
@@ -0,0 +1,93 @@
+#ifndef FWCore_Framework_interface_GenericProduct_h
+#define FWCore_Framework_interface_GenericProduct_h
+
+#include <cstring>
+#include <memory>
+#include <typeinfo>
+#include <utility>
+
+#include "DataFormats/Common/interface/OrphanHandle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Reflection/interface/ObjectWithDict.h"
+#include "FWCore/Reflection/interface/TypeWithDict.h"
+
+namespace edm {
+
+  class GenericProduct {
+  public:
+    // TODO make private and add accessors and a constructor
+    TypeWithDict wrappedType_;
+    ObjectWithDict object_;
+  };
+
+  template <>
+  class OrphanHandle<GenericProduct> : public OrphanHandleBase {
+  public:
+    OrphanHandle(ObjectWithDict prod, ProductID const& id) : OrphanHandleBase(prod.address(), id), prod_(prod) {}
+
+    ObjectWithDict const* product() const { return &prod_; }
+
+    ObjectWithDict const* operator->() const { return &prod_; }
+
+    ObjectWithDict const& operator*() const { return prod_; }
+
+  private:
+    ObjectWithDict prod_;
+  };
+
+  // specialise Event::putImpl for GenericProduct
+  template <>
+  inline OrphanHandle<GenericProduct> Event::putImpl(EDPutToken::value_type index,
+                                                     std::unique_ptr<GenericProduct> product) {
+    static const TypeWithDict classDoNotSortUponInsertion(typeid(DoNotSortUponInsertion));
+    static const TypeWithDict classWrapperBase(typeid(WrapperBase));
+
+    // call post_insert only if the underlying type has such a function
+    // FIXME functionMemberByName causes ROOT to parse the headers, which we generally want to avoid.
```

Comment:
I'm not sure how to handle this.

Maybe add another virtual method to `WrapperBase`, and specialise it in `Wrapper<T>` depending if `T` has a `post_insert` method ?